### ISSUE TITLE
Issue uploading to pypi with supplied versioning scheme

### ISSUE THREAD

**Issue by jtmorrisbytes** [zsimic/setupmeta#8]
Hi zsimic, I love using your setupmeta package, but I think the verioning system is out of date. I am getting errors uploading to Pypi. while deploying to travis using post, build-id or '{major}.{minor}.{patch}' I get an error that says "400 Client Error: version: Cannot use PEP 440 local versions. for url: https://upload.pypi.org/legacy/" Is there anything that I can do to keep using the code that already exists?

**Comment by zsimic** [comment]
Can you point me to your project? I wonder if it's a versioning strategy choice, as I don't have any trouble publishing to pypi or elsewhere.

I'll gladly take a look, point me to your repo if you can, or send me the relevant part of your `setup.py`

**Comment by jtmorrisbytes** [comment]
https://github.com/jtmorrisbytes/pyjtmorrisbytes
I want to try to use src/pyjtmorrisbytes/__init__.py to store the pertinant info

**Comment by zsimic** [comment]
The `setup()` call in `setup.py` has no `name=` specified, that's the starting point usually.

Also, you are specifying `version='0.1.5' ` explicitly and no `versioning=...` (for setupmeta), which means that setupmeta will not try and determine the version of this project (as a rule, setupmeta will never touch anything you explicitly state in `setup()` call).

I'd say you're not really using setupmeta here, it is part of your `setup_requires` list, but it won't get to do much, especially since you seem to also specify your own `cmdclass`... and a `setuptools.installations` :)

I would start from an as-simple-as-possible `setup.py` first, current one seems to mix a lot of stuff.

One simple thing you could try is replace temporarily your `setup.py` with this:
```
from setuptools import setup

setup(
    name="pyjtmorrisbytes",
    versioning="dev",
    setup_requires=['setupmeta'],
)
```

And, delete (or empty out) `setup.cfg`.

Now, try running `python setup.py explain`, and you should get something like this:
```
                       author: (src/pyjtmorrisbytes/__init__.py:13) Jordan Taylor Morris
                  description: (README.md:1                       ) flask-pyjtmorrisbytes-api
                       email*: (src/pyjtmorrisbytes/__init__.py:14) jthecybertinkerer@gmail.com
...
             long_description: (README.md                         ) # flask-pyjtmorrisbytes-api the flask version of my api
long_description_content_type: (README.md                         ) text/markdown
                         name: (explicit                          ) pyjtmorrisbytes
                  package_dir: (auto-fill                         ) {: src}
                     packages: (auto-fill                         ) 6 items: ["pyjtmorrisbytes", "pyjtmorrisbytes.config", ...
...
                      version: (git                               ) 0.1.1.dev6+g44cfe56
                           \_: (src/pyjtmorrisbytes/__init__.py:7 ) 0.1.5
                   versioning: (explicit                          ) dev
                           \_: (src/pyjtmorrisbytes/__init__.py:8 ) branch(dev,master):{major}.{minor}.{patch}
```

This would show you what setupmeta managed to find from your project. 
From there, you could start adding stuff and see which parts become incompatible.

setupmeta was designed for very simple `setup.py`-s, not complex ones.
It should be able to handle stuff like `.pyx` extensions etc without issue, but I don't think having a lot of `setup_requires` will be a good use-case for setupmeta, like this line:
```
 setup_requires=['pytest-runner', 'cython', 'setupmeta', 'tox-setuptools', 'sphinx'],
```

That is mixing quite a few things:
- usually you don't want to require `pytest-runner`, `tox-setuptools` and `sphinx` at setup time, that's much too early, you want that much later, during test runs and doc generation
- don't know what `cython` does at setup time

**Comment by jtmorrisbytes** [comment]
Ill give your suggestion a try. My project is a cython based project that uses compiled pyx files. I need to either call Cython.Build.Cythonize with a list of pyx files or use the command line with a list of files to compile.

I am also having trouble figuring out how to prepare such a project for distribution. In its current state, It could compile files and upload them to travis, which then could upload a tar.gz file, but then all installations fail.  Im afraid that I am severely misunderstanding the packaging and setup process, and have had to mangle a bunch of different tutorials together to get anything to work

**Comment by jtmorrisbytes** [comment]
After testing your suggestion, I can confirm setup.py version --bump patch --commit works as expected. I have also confrmed that setupmeta detects package information correctly. I Think I have to add a MANIFEST.in file to include .pyx, .c, .so, and .pyd files. does setupmeta detect those files for sdist, bdist, and bdist_wheel? it would be greate if they could so I didn't have to use the manifest file.

Just tried to run tests, and pytest isnt picking up any tests without specifying anything in setup.py

**Comment by jpic** [comment]
MANIFEST.in is not necessary / out of scope of setupmeta. However, you will find this app pretty helpful: https://pypi.org/project/check-manifest/

**Comment by zsimic** [comment]
Packaging is hard indeed :) setupmeta tries to help with the metadata part, making it less repetitive to have most of the common things auto-populated from other parts that you supposedly already have in your project (like long description from README, which is very standard practice)

I am not very familiar with pyx nor MANIFEST.in I must admit.

**Comment by zsimic** [comment]
Regarding tests, note that setupmeta will declare them properly if you do this:

```
mv tests/jtmorrisbytes/* tests/
rm -r tests/jtmorrisbytes
cat << EOF > tests/requirements.txt
pytest
pytest-cython
EOF
```

Now run `python setup.py explain`, and you should see this line:
```
  tests_require: (tests/requirements.txt ) ["pytest", "pytest-cython"]
```

(make sure to NOT declare `tests_require` explicitly in your `setup.py` in this case, to let setupmeta find it).
Basically you have your tests in non-standard location `tests/jtmorrisbytes`, which setupmeta does not look at.

See for example this: https://github.com/requests/requests/tree/master/tests (pretty standard: tests are all right under `tests/` folder)

**Comment by zsimic** [comment]
Regarding author, setuptools has that weird historical thing where they want 2 fields... setupmeta is trying to let you do just one (and it will try and reasonably make the expected 2 appropriately).

In your `__init__.py`, try changing the 2 corresponding lines to this (there was a typo in license):
```
__license__ = 'MIT'
__author__ = "Jordan Taylor Morris jthecybertinkerer@gmail.com"
```

With this, you should see `python setup.py explain` now show:
```
       author: (auto-adjust                       ) Jordan Taylor Morris
           \_: (src/pyjtmorrisbytes/__init__.py:13) Jordan Taylor Morris jthecybertinkerer@gmail.com
 author_email: (auto-adjust                       ) jthecybertinkerer@gmail.com
      license: (src/pyjtmorrisbytes/__init__.py:9 ) MIT
```

This should allow you to simplify the `from src.pyjtmorrisbytes import __email__` from your `setup.py`.

Note that once everything is configured properly, you should never have to mention `src` from any import.

Using `src/` subfolder is usually good practice (but debated amongst pythonistas...), I find it the more powerful way to go.
It tucks all your code neatly under one folder, but it does make it an extra challenge to package your project correctly for it to work. Usually people against `src/` folder don't want to bother making their package 100% correct (or they have a reason to want "shortcuts" for certain local iterations...)

The `src/` thing would be a long topic, but in general, if packaged right, you never ever mention `src` in any import, yet your packaging setup is such that `import <mylibname>` works correctly.
Using `src/` also forces you to always run your code from a venv (except `setup.py`), which is usually very good practice.

**Comment by zsimic** [comment]
> MANIFEST.in is not necessary / out of scope of setupmeta. However, you will find this app pretty helpful: https://pypi.org/project/check-manifest/

Yup, setupmeta doesn't look or know much about MANIFEST.in yet (probably because I am also not familiar with it :) so no one taught it yet about that file)

Note you can run a tool such as `check-manifest` from tox, a little bit like this https://github.com/zsimic/setupmeta/blob/master/tox.ini#L30-L36 (`readme_renderer` in that example, but same idea)

**Comment by jtmorrisbytes** [comment]
Got it. I employed your sugestions without a problem. Im just waiting on a stable build process before attempting to upload to pypi again

**Comment by jtmorrisbytes** [comment]
I can successfully upload to pypi using setuptools and setupmeta, so I will consider this issue closed
